### DISCUSSION TITLE
affine extend option doesn't have any effect

### DISCUSSION THREAD

**Discussion by stoodfarback** [libvips/libvips#4259] [Q&A]
Using this example image as in.jpg:
![in](https://github.com/user-attachments/assets/973e3bb7-e7d5-4d95-b466-5ef140b65d48)

**Rotate by 45 degrees and use extend "mirror" option, new pixels are black instead:**
`> vips affine in.jpg out.jpg "0.7071067811865476 0.7071067811865475 -0.7071067811865475 0.7071067811865476" --extend mirror`
![out](https://github.com/user-attachments/assets/2d6d920c-0e75-4d1f-89a2-8a1d638404a5)

**The background option works as expected:**
`> vips affine in.jpg out2.jpg "0.7071067811865476 0.7071067811865475 -0.7071067811865475 0.7071067811865476" --background 255`
![out2](https://github.com/user-attachments/assets/2eef2c0a-601d-464c-9fc1-fe2165a8698d)

**Expected:**
`> magick in.jpg -virtual-pixel mirror +distort SRT -45 out3.jpg`
![out3](https://github.com/user-attachments/assets/1e88a544-b9c3-4a2e-a142-a846913c7513)

**Using extend "white" option shows odd behaviour, adding a small white border:**
`> vips affine in.jpg out4.jpg "0.7071067811865476 0.7071067811865475 -0.7071067811865475 0.7071067811865476" --extend white`
![out4](https://github.com/user-attachments/assets/a43a070a-2b40-4122-a1bb-19da359307da)

**Note**
Ideally, I'd prefer to use the `rotate` function instead of `affine`, but it doesn't support the extend option:
`> vips rotate in.jpg out5.jpg 45 --extend mirror`
`Unknown option --extend`

**Environment**
 - OS: Arch Linux (6.11.6 kernel)
 - Vips: 8.16.0

**Comment by jcupitt** [comment]
Hi @stoodfarback,

The `--extend` option controls the antialiasing of edge pixels, not the background. If you want to mirror the whole of the background, use `gravity` beforehand and `crop` after.

This isn't very efficient at the CLI, since it will generate much larger images, but it's quick in something like python since libvips will only compute the pixels it needs for the final output.

**Comment by stoodfarback** [comment]
Thank you for the response. Maybe worth updating the docs to reflect this?

Current `vips_affine` description:
`extend : VipsExtend how to generate new pixels`
and:
`By default, new pixels are filled with background. This defaults to zero (black). You can set other extend types with extend.`
And the `VipsExtend` description:
`When the edges of an image are extended, you can specify how you want the extension done.`

Part of my confusion and the reason I thought this was a bug is that vips_embed uses the same `VipsExtend` enum and works as I was expecting:
`> vips embed in.jpg out6.jpg 50 50 300 300 --extend mirror`
![out6](https://github.com/user-attachments/assets/6b4d9c64-dfbd-4e8e-bc7c-08fd05917838)
